### PR TITLE
fix #6896

### LINKED ISSUE #6896: Add support for Sticky Bit permissions in OS module
As per title, could support for the Sticky Bit (S_ISVTX) be added?

Quick patch that passes stdlib test suite:
[Patch.diff](https://pastebin.com/PpekPnrG)
[Patch.test](https://pastebin.com/6bFNFUeH)
Program output with patch:
[Patch.ls](https://pastebin.com/Cc24ft9A)

Thank you

### PR DESCRIPTION
**Pull request by NopeDK**
First attempt at using Git properly, be gentle ;)

### CODE REVIEW

File: lib/pure/os.nim [hunk 1]
```diff
@@ -447,6 +448,8 @@ proc getFilePermissions*(filename: string): set[FilePermission] {.
     if (a.st_mode and S_IROTH) != 0'i32: result.incl(fpOthersRead)
     if (a.st_mode and S_IWOTH) != 0'i32: result.incl(fpOthersWrite)
     if (a.st_mode and S_IXOTH) != 0'i32: result.incl(fpOthersExec)
+
```
Comment (Yardanico) [inline lib/pure/os.nim]: Maybe remove this space to be consistent with other if's?

Comment (NopeDK) [reply lib/pure/os.nim]: To me it looks like the permissionsare grouped by User, Group and Other. I don't know if Sticky Bit fits into those categories so I added the newline as if it were in a separate category. Same with the checkandincludemode further down.

Comment (Yardanico) [reply lib/pure/os.nim]: Ah, sorry, didn't see the whole thing

**Comment by NopeDK** [comment]
Failure to build found. Working to fix now.

File: lib/pure/asyncftpclient.nim [hunk 2]
```diff
@@ -152,7 +152,7 @@ proc connect*(ftp: AsyncFtpClient) {.async.} =
   if ftp.pass != "":
     assertReply(await(ftp.send("PASS " & ftp.pass)), "230")
 
-proc pwd*(ftp: AsyncFtpClient): Future[TaintedString] {.async.} =
+roc pwd*(ftp: AsyncFtpClient): Future[TaintedString] {.async.} =
```
Comment (Yardanico) [inline lib/pure/asyncftpclient.nim]: typo :) "roc"

Comment (NopeDK) [reply lib/pure/asyncftpclient.nim]: Damn keyboard. Fixed on newer commit.

**Comment by NopeDK** [comment]
@dom96 Windows disregards everything but read-only. If set to read-only, the fp*Write's are removed, otherwise all permissions are set.

**Comment by NopeDK** [comment]
Checks passed, ready for merge.

**Comment by dom96** [comment]
> @dom96 Windows disregards everything but read-only. If set to read-only, the fp*Write's are removed, otherwise all permissions are set.

That seems incorrect, we should probably change that in the future.

File: lib/pure/os.nim [hunk 3]
```diff
@@ -423,7 +423,8 @@ type
     fpGroupRead,           ## read access for the group
     fpOthersExec,          ## execute access for others
     fpOthersWrite,         ## write access for others
-    fpOthersRead           ## read access for others
+    fpOthersRead,          ## read access for others
+    fpStickyBit
```
Comment (Araq) [inline lib/pure/os.nim]: What is the "sticky" bit? Documentation comment missing.

**Comment by Araq** [comment]
-1 from me. ``os.nim`` is an abstration over the Posix and Windows APIs or other OS specific APIs and as such is more a most common denominator than a "everything that any OS ever supported" kitchen sink. Read/write/exec access rights make sense conceptually, "sticky" seems like a random hack somebody invented 4 decades ago. If you need to access Posix features, use ``posix.nim``.

**Comment by Araq** [comment]
> That seems incorrect, we should probably change that in the future.

Seems exactly the right behaviour to me.

**Comment by dom96** [comment]
> os.nim is an abstration over the Posix and Windows APIs or other OS specific APIs and as such is more a most common denominator than a "everything that any OS ever supported" kitchen sink.

The sticky bit is supported by all Unix-like OS', because of that I think supporting it is worthwhile. Windows is the exception here. Also, even the doc comment for ``FilePermission`` states "file access permission; modelled after UNIX"

**Comment by Araq** [comment]
> The sticky bit is supported by all Unix-like OS', because of that I think supporting it is worthwhile. 

And what does it **mean** on all Unix-like OSes?

**Comment by NopeDK** [comment]
The problem with not having the sticky bit is the following:
`assert((getFilePermissions("non_sticky_folder") == getFilePermissions("sticky_folder"))`
will succeed but the actual permissions:
```
import posix
var non_sticky, sticky: Stat
discard stat("non_sticky_folder", non_sticky)
discard stat("sticky_folder", sticky)
assert(non_sticky.st_mode == sticky.st_mode)
```
will fail. (non_sticky.st_mode = 16895, sticky.st_mode = 17407

The lack of sticky bit also translates to copyDirWithPermissions where the previous will happen as it relies on getFilePermissions and setFilePermissions that do not include the sticky bit.
The bit also has multiple meanings and use cases as seen in the table on [Wiki/Sticky_bit#Usage](https://en.wikipedia.org/wiki/Sticky_bit#Usage)

**Comment by NopeDK** [comment]
For full permissions, S_ISUID and S_ISGID should be added as well per [man inode(7)](http://man7.org/linux/man-pages/man7/inode.7.html).
@Araq

**Comment by dom96** [comment]
> This is not gonna make it into my precious os.nim, sorry.

How does including this hurt your precious os module? :)

**Comment by Araq** [comment]
> For full permissions, S_ISUID and S_ISGID should be added as well per man inode(7).
Though I don't know how common those are compared to the fairly standard S_ISVTX.

So, in other words, we made the data model uglier without fixing the bug.

**Comment by Araq** [comment]
> The lack of sticky bit also translates to copyDirWithPermissions where the previous will happen as it relies on getFilePermissions and setFilePermissions that do not include the sticky bit.

copyDirWithPermissions could use stat() instead but IMO that is not a bug. Permissions are what is declared by ``os.nim`` and nothing else needs to be taken care of by copyDirWithPermissions. You might as well argue that file timestamps are not copied over. Oh and it doesn't care about Linux's ACL either...

**Comment by NopeDK** [comment]
Normally a copy preserves nothing, when you have copyWithPermissions then you state that permissions are copied.
Timestamps are not copied normally unless stated so not preserving those is in line with the current function.
ACL can be discussed as this seems to be advanced Unix permissions which is more in line with the posix module.
I feel that permissions should at least include all the basic octals, which include the S_ISGID, S_ISUID and S_ISVTX flags.
I will continue the work on this branch during this week to reflect what I have written above; summarized:
Adding the two flags
Stating that timestamps are not copied in the docs for relevant procs
And that ACL is beyond the scope of os.nim and can be found  in posix.nim.

**Comment by Araq** [comment]
This is still a breaking change, now every time I use the FilePermission enum I have to consider archaic bits with no consistent meaning between OSes. Still -1 for me. Just fix the implementation of copyDirWithPermissions and give the FTP client an overload to accept an ``uint16`` for the file permissions or something like that.

**Comment by NopeDK** [comment]
I can see your point.
I feel rewriting the copyFileWithPermissions and copyDirWithPermissions to handle this, without changing the enum, will get too hacky so I will go ahead and close this.

**Pull request closed without merging**